WordPress/openverse-catalog
# Elasticsearch metrics DAG

## Fixes


Fixes #564 by @AetherUnbound

## Description


This PR adds a DAG which captures Elasticsearch metrics for the 72 hours prior to the schedule interval and reports them to slack.

I also took this opportunity to use the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html) for Airflow - it made the actual DAG definition so much simpler! Super useful when all of your steps would otherwise be `PythonOperators`. 

### Additional notes
- This adds a few new variables/connections. With the way this is set up though, nothing (should) need to be modified in production besides adding the `ES_INSTANCE_IDS` Variable.
- This adds a new python package ([PIL](https://pillow.readthedocs.io/en/stable/)) in order to perform some image manipulation. This means we'll need to publish and deploy a new Docker image in order to actually run this DAG. We could do this as a specific commit, or we could cut and release v1.3.1.
- This adds a new bucket to S3, `openverse-airflow`. This bucket can be used for Airflow entities that don't explicitly relate to the catalog (since `openverse-catalog` almost exclusively has TSVs in it). I've added this bucket by hand for testing, but I recognize that it should probably have a follow-up in the infrastructure repo. @sarayourfriend do you have any insight on where it'd be best to put that?
- I tried to make the DAG agnostic to the number of metrics we're gathering, in case we want to add more in the immediate future.

### Screenshots
**DAG**
![image](https://user-images.githubusercontent.com/10214785/179115806-9a624617-d508-4878-9a7a-5f1895ad6da4.png)


**Slack Message**
![Screenshot_2022-07-14_15-48-42](https://user-images.githubusercontent.com/10214785/179115830-4827a0e3-b00d-4737-b5df-411d4232a930.png)


## Testing Instructions


In order to test this DAG locally, you will need to set the following variables in your `.env` file:
```bash
AIRFLOW_VAR_ES_INSTANCE_IDS=i-00e708a227372b51b,i-0e609c909948a7a02,i-075774d799a0da4b3
AIRFLOW_CONN_AWS_METRICS=aws://<your-key>:<your-secret-url-encoded>@?region_name=us-east-1
AWS_METRICS_CONN_ID=aws_metrics
```

**Note**: The metrics connection must use your actual AWS credentials, otherwise the DAG will not be able to retrieve metrics from CloudWatch. The secret key **must** be URL encoded as well, you can do so with this command:

```bash
python -c 'import urllib.parse; safe_string = urllib.parse.quote_plus("<your-secret>"); print(safe_string)'
```

Once this is set up, run `just up` and turn on the DAG! The DAG should kick off a run on its own. This will generate a PNG from the metrics and upload it to your local Minio instance. You can view the generated PNG once the DAG is complete by navigating to http://localhost5011 and going to `openverse-airflow/metrics/elasticsearch/<date>/es-metrics.png` and then click "Preview".

![Screenshot_2022-07-18_14-40-27](https://user-images.githubusercontent.com/10214785/179622463-b2dfae55-e770-442d-9842-142070e9b52a.png)


## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
